File: 2015/c/2015-02.c
#include <stdio.h>
#include <string.h>

#include "tools.h"

void sort(int *arr) {
  int swapped;
  
  while (1) {
    swapped = 0;

    for (int i = 0; i < 2; i++) {
      if (arr[i] > arr[i + 1]) {
        int tmp = arr[i];
        arr[i] = arr[i + 1];
        arr[i + 1] = tmp;
        swapped = 1;
      } 
    }

    if (swapped == 0) break;
  }
}


int part1(char input[]) {
  int output = 0;
  char data[strlen(input)];

  strcpy(data, input);

  char *endLine;
  char *line = strtok_r(data, "\n", &endLine);

  while (line != NULL) {
    int l, w, h;
    int sides[3];

    l = atoi(strtok(line, "x"));
    w = atoi(strtok(NULL, "x"));
    h = atoi(strtok(NULL, "x"));

    sides[0] = l * w;
    sides[1] = w * h;
    sides[2] = h * l;

    sort(sides);

    output += 2 * sides[0];
    output += 2 * sides[1];
    output += 2 * sides[2];
    output += sides[0];

    line = strtok_r(NULL, "\n", &endLine);
  }

  return output;
}

int part2(char input[]) {
  int output = 0;
  char data[strlen(input)];

  strcpy(data, input);

  char *endLine;
  char *line = strtok_r(data, "\n", &endLine);

  while (line != NULL) {
    int dims[3];

    dims[0] = atoi(strtok(line, "x"));
    dims[1] = atoi(strtok(NULL, "x"));
    dims[2] = atoi(strtok(NULL, "x"));

    sort(dims);

    output += dims[0] + dims[0];
    output += dims[1] + dims[1];
    output += dims[0] * dims[1] * dims[2];

    line = strtok_r(NULL, "\n", &endLine);
  }

  return output;
}

int main() {
  char *data = getDataFromFile("../data/2015-02.txt");

  printf("Advent of code 2015 #02\n\n");

  printf("----Part 1----\n");
  test("Test 1", part1("2x3x4"),  58);
  test("Test 2", part1("1x1x10"), 43);
  printf("Answer: %d\n\n", part1(data));

  printf("----Part 2----\n");
  test("Test 1", part2("2x3x4"),  34);
  test("Test 2", part2("1x1x10"), 14);
  printf("Answer: %d\n\n", part2(data));

  return 0;
}

File: 2015/js/2015-02.js
const path  = require('path')
const fs    = require('fs')

const data = fs.readFileSync(path.resolve(__dirname, '../data/2015-02.txt'), 'utf8')

function part1(input) {
  let output = 0
  const data = input.split('\n')

  data.forEach(dim => {
    const [ l, w, h ] = dim.split('x')
    const sides = [l*w, w*h, h*l]
    
    sides.sort((a, b) => a - b)

    output += 2 * sides[0]
    output += 2 * sides[1]
    output += 2 * sides[2]
    output += sides[0]
  })

  return output
}

function part2(input) {
  let output = 0
  const data = input.split('\n')

  data.forEach(box => {
    const dims = box.split('x')
    dims.sort((a, b) => a - b)

    output += parseInt(dims[0]) + parseInt(dims[0])
    output += parseInt(dims[1]) + parseInt(dims[1])
    output += dims[0]*dims[1]*dims[2]
  })

  return output
}

console.log( 'Part 1 tests' )
console.assert( part1('2x3x4')  == 58 )
console.assert( part1('1x1x10') == 43 )

console.log( 'Part 2 tests' )
console.assert( part2('2x3x4')  == 34 )
console.assert( part2('1x1x10') == 14 )

console.log( `
  Advent of code 2015 #02
  Part1: ${part1(data)}
  Part2: ${part2(data)}
` )